design/ceph/osd-migration.md

```diff
@@ -0,0 +1,239 @@
+  2. If the PG status is not `active+clean`, don't replace any OSD.
+  3. Once the PG status is `active+clean`:
+   - If `OSDInProgress` is not empty due to any error in the previous reconcile, then don't remove any new OSD from the `OSDReplaceConfig.OSDs` and continue with the deletion of the OSD mentioned in `OSDInProgress`.
+   - Else remove an OSD entry from the `OSDReplaceConfig.OSDs` and assign it to `OSDInProgress`
+  4. Provide a new environment variables to the OSD prepare pod running on `OSDReplaceInfo.Node` in case of OSD on disks or `OSDReplaceInfo.Path` in case of OSD on PVCs
+      - `ROOK_REPLACE_OSD_ID`: OSD to be deleted.
+  5. OSD prepare pod job will destroy the OSD based on the environment variables provided above. This process is mentioned in the `Destroy OSD` steps below.
+  6. Continue with the existing flow of creating and upgrading the OSD, with the exception of skipping the upgrade for all the OSDs that are currently present in the `OSDReplaceConfig.OSDs`. Skipping update operation on the OSDs that need to be migrated is important because with OSDs on disks, the OSD prepared pod is updated if there is an update in the spec. This will update the osd store to `bluestore-rdr` in the spec without migrating the OSD. This will cause errors.
+  7. Once the destroy OSD pod is added back, update `LastReplacedOSD` with `OSDInProgress` and `LastReplacedTimeStamp` with current time. Reset `OSDInProgress` to nil. If there is any error during the OSD migration, then don't reset `OSDInProgress`.
+  8. Reconcile the operator and perform as same steps in the next reconcile.
+
+#### Destroy OSD
+
+OSD prepare pod job will perform OSD deletion using following steps.
+- It checks for `ROOK_REPLACE_OSD_ID` arguments.
+- Use `ceph volume list` to fetch the OSD path.
+- Delete the OSD deployment based on the `ROOK_REPLACE_OSD_ID`
+- Destroy the OSD based on the `ROOK_REPLACE_OSD_ID`
+
+    ```
+    ceph osd destroy {ROOK_REPLACE_OSD_ID} --yes-i-really-mean-it
+    ```
+
+- Once the OSD is destroyed, wipe the backing device. This removes all the data on the device.
+- Once the device is wiped, prepare the OSD with the new store type. Use the same OSD ID to prepare the OSD again. This is done by passing the OSD ID as `--osd-id` flag to `ceph-volume`.
+- Prepare the OSD using the same `ROOK_REPLACE_OSD_ID` and path using following commands:
+
+    ```
+    ceph-volume lvm prepare --osd-id {ROOK_REPLACE_OSD_ID} --data {OSD_PATH}
+    ```
+
+#### Manage OSD migration status
+
+- OSD migration status will managed by `OSDReplaceConfig` API. This will be stored as the config map. For example:
+``` yaml
+apiVersion: v1
+data:
+  config: '{"osdList":{"osds":[{"id":1,"path":"/mnt/set1-data-1qwdpk","node":"minikube-m03"},{"id":2,"path":"/mnt/set1-data-2c8h9k","node":"minikube-m04"}]},"isPgHealthy":true,"pgHealthMessage":"all
```

Comment:
It still feels like we only need two items in this configmap: list of OSD IDs still to migrate, and the osdInProgress

Reply:
removed `LastReplacedOSD` and `LastReplacedOSDTimestamp` and `PGHealthMessage` from the config.